File: pylith/meshio/OutputSolnSubset.py
# ----------------------------------------------------------------------
#
# Brad T. Aagaard, U.S. Geological Survey
# Charles A. Williams, GNS Science
# Matthew G. Knepley, University of Chicago
#
# This code was developed as part of the Computational Infrastructure
# for Geodynamics (http://geodynamics.org).
#
# Copyright (c) 2010-2017 University of California, Davis
#
# See COPYING for license information.
#
# ----------------------------------------------------------------------
#
# @file pyre/meshio/OutputSolnSubset.py
#
# @brief Python object for managing output of finite-element solution
# information over a subdomain.
#
# Factory: observer

from .OutputManager import OutputManager
from .meshio import OutputSolnSubset as ModuleOutputSolnSubset


def validateLabel(value):
    """
    Validate label for group/nodeset/pset.
    """
    if 0 == len(value):
        raise ValueError("Label for group/nodeset/pset in mesh not specified.")
    return value


class OutputSolnSubset(OutputManager, ModuleOutputSolnSubset):
    """
    Python object for managing output of finite-element solution
    information over a subdomain.

    INVENTORY

    Properties
      - *label* Name identifier for subdomain.

    Facilities
      - None

    Factory: output_manager
    """

    import pyre.inventory

    label = pyre.inventory.str("label", default="", validator=validateLabel)
    label.meta['tip'] = "Label identifier for subdomain."

    # PUBLIC METHODS /////////////////////////////////////////////////////

    def __init__(self, name="outputsolnsubset"):
        """
        Constructor.
        """
        OutputManager.__init__(self, name)
        return

    def preinitialize(self):
        """
        Do
        """
        OutputManager.preinitialize(self)
        ModuleOutputSolnSubset.label(self, self.label)
        return

    def verifyConfiguration(self, mesh):
        """
        Verify compatibility of configuration.
        """
        OutputManager.verifyConfiguration(self, mesh)
        ModuleOutputSolnSubset.verifyConfiguration(self, mesh)
        return

    # PRIVATE METHODS ////////////////////////////////////////////////////

    def _configure(self):
        """
        Set members based using inventory.
        """
        OutputManager._configure(self)
        return

    def _createModuleObj(self):
        """
        Create handle to C++ object.
        """
        ModuleOutputSolnSubset.__init__(self)
        return


# FACTORIES ////////////////////////////////////////////////////////////

def observer():
    """
    Factory associated with OutputManager.
    """
    return OutputSolnSubset()


# End of file

File: libsrc/pylith/meshio/OutputSolnSubset.cc
// -*- C++ -*-
//
// ======================================================================
//
// Brad T. Aagaard, U.S. Geological Survey
// Charles A. Williams, GNS Science
// Matthew G. Knepley, University of Chicago
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010-2015 University of California, Davis
//
// See COPYING for license information.
//
// ======================================================================
//

#include <portinfo>

#include "OutputSolnSubset.hh" // implementation of class methods

#include "pylith/topology/Mesh.hh" // USES Mesh

// ----------------------------------------------------------------------
// Constructor
pylith::meshio::OutputSolnSubset::OutputSolnSubset(void) :
  _label(""),
  _submesh(0)
{ // constructor
} // constructor

// ----------------------------------------------------------------------
// Destructor
pylith::meshio::OutputSolnSubset::~OutputSolnSubset(void)
{ // destructor
  deallocate();
} // destructor  

// ----------------------------------------------------------------------
// Deallocate PETSc and local data structures.
void
pylith::meshio::OutputSolnSubset::deallocate(void)
{ // deallocate
  PYLITH_METHOD_BEGIN;

  OutputManager::deallocate();

  delete _submesh; _submesh = 0;

  PYLITH_METHOD_END;
} // deallocate
  
// ----------------------------------------------------------------------
// Set label identifier for subdomain.
void
pylith::meshio::OutputSolnSubset::label(const char* value)
{ // label
  PYLITH_METHOD_BEGIN;

  _label = value;

  PYLITH_METHOD_END;
} // label

// ----------------------------------------------------------------------
// Verify configuration is acceptable.
void
pylith::meshio::OutputSolnSubset::verifyConfiguration(const topology::Mesh& mesh) const
{ // verifyConfiguration
  PYLITH_METHOD_BEGIN;

  PetscDM dmMesh = mesh.dmMesh();assert(dmMesh);
  PetscBool hasLabel = PETSC_FALSE;
  PetscErrorCode err = DMHasLabel(dmMesh, _label.c_str(), &hasLabel);PYLITH_CHECK_ERROR(err);
  if (!hasLabel) {
    std::ostringstream msg;
    msg << "Mesh missing group of vertices '" << _label
	<< " for subdomain output.";
    throw std::runtime_error(msg.str());
  } // if

  PYLITH_METHOD_END;
} // verifyConfiguration

// ----------------------------------------------------------------------
// Get mesh associated with subdomain.
const pylith::topology::Mesh&
pylith::meshio::OutputSolnSubset::subdomainMesh(const topology::Mesh& mesh)
{ // subdomainMesh
  PYLITH_METHOD_BEGIN;

  delete _submesh; _submesh = new topology::Mesh(mesh, _label.c_str());assert(_submesh);

  PYLITH_METHOD_RETURN(*_submesh);
} // subdomainMesh


// End of file 
